tools/isobuild/package-api.js

```diff
@@ -339,13 +339,39 @@ _.extend(PackageAPI.prototype, {
     var errors = [];
     _.each(paths, function (path) {
       forAllMatchingArchs(arch, function (a) {
+        const isAsset = !! (fileOptions && fileOptions.isAsset);
+
+        // Check if we have already added a file at this path
         if (_.has(self.sources[a], path)) {
-          errors.push("Duplicate source file: " + path);
+          const otherSource = self.sources[a][path];
+
+          // Files can actually be added twice for each architecture: once as
+          // an asset, and once as a source file
+          const otherFileOptions = otherSource.fileOptions;
+          const otherFileIsAsset = !! (otherFileOptions &&
```

Comment:
You're right that this code is a bit janky, and I totally missed the case where `isAsset: "both"` and you try to add the file a third time. It should definitely print a helpful error.

However, I don't think having `isAsset: "both"` as an external API will work right out of the box, because you will want to control on which architectures it is a source file and on which it is an asset. We could add additional options to work around that, but it could be more work and adding the file twice seems like an OK API to me. What do you think?

Another option instead of having the "both" special flag is to change the data structure for sources to have more levels: arch -> type -> path, where type is "asset" or "source". That would make the code easier to reason about, but could involve more changes to compiler.